File: js/src/algo_gen/promote.js
/**
 * promote.js — Registry promotion for validated algorithms.
 *
 * Mirrors python/algoforge/algo_gen/promote.py.
 *
 * Public API:
 *   promote(report, manifest, registryDir, opts?) -> ManifestAlgo
 */

import { writeFileSync, mkdirSync } from 'node:fs';
import { join } from 'node:path';
import { ManifestAlgo } from './runtime.js';
import { TierName } from './types.js';

// Tier color codes (mirrors Python _TIER_COLORS)
const TIER_COLORS = {
    [TierName.red]:    '#ff4444',
    [TierName.orange]: '#ff8800',
    [TierName.yellow]: '#ffdd00',
    [TierName.green]:  '#00ff88',
    [TierName.white]:  '#ffffff',
};

export class PromotionError extends Error {
    constructor(msg) { super(msg); this.name = 'PromotionError'; }
}

/**
 * Promote a validated manifest into the registry.
 * Refuses red tier. Returns a ManifestAlgo with .metadata populated.
 *
 * @param {import('./types.js').TierReport} report
 * @param {import('./types.js').AlgoManifest} manifest
 * @param {string} registryDir
 * @param {object} [opts]
 * @param {object|null} [opts.cfg]              tier config overrides
 * @param {object|null} [opts.generatorInfo]    generator metadata
 * @param {string}      [opts.brief]            brief description
 * @param {boolean}     [opts.confirm]          required for orange/yellow
 * @returns {ManifestAlgo}
 */
export function promote(report, manifest, registryDir, opts) {
    const cfg           = (opts && opts.cfg)           ? opts.cfg           : null;
    const generatorInfo = (opts && opts.generatorInfo)  ? opts.generatorInfo  : {};
    const brief         = (opts && opts.brief)          ? opts.brief          : '';
    const confirm       = (opts && opts.confirm)        ? opts.confirm        : false;

    // Refuse red tier
    if (report.tier === TierName.red) {
        throw new PromotionError(
            `Cannot promote algorithm '${manifest.name}': ` +
            `tier is red (min_score=${report.minScore.toFixed(2)} < 70). ` +
            `Red-tier algorithms are never promoted.`
        );
    }

    // Check manual review gate (orange/yellow require confirm=true)
    const requireBelow = (cfg && cfg.require_manual_review_below) || 'green';
    const tiersRequiringConfirm = new Set();
    if (requireBelow === 'green' || requireBelow === 'white') {
        tiersRequiringConfirm.add(TierName.orange);
        tiersRequiringConfirm.add(TierName.yellow);
    } else if (requireBelow === 'yellow') {
        tiersRequiringConfirm.add(TierName.orange);
    }

    if (tiersRequiringConfirm.has(report.tier) && !confirm) {
        throw new PromotionError(
            `Cannot promote '${manifest.name}' (tier=${report.tier}) ` +
            `without confirm=true. This tier requires manual review.`
        );
    }

    // Build metadata dict per spec §7
    const createdAt    = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
    mkdirSync(registryDir, { recursive: true });
    const manifestPath = join(registryDir, `${manifest.name}.json`);

    const metadata = {
        source:         'algo_gen',
        schema_version: manifest.schemaVersion,
        tier:           report.tier,
        tier_color:     TIER_COLORS[report.tier] || '#888888',
        scores: {
            walk_forward: report.walkForward,
            mc_bootstrap: report.mcBootstrap,
            robustness:   report.robustness,
        },
        min_score:     report.minScore,
        size_mult:     report.sizeMult,
        experimental:  report.experimental,
        paper_only:    report.paperOnly,
        created_at:    createdAt,
        generator:     generatorInfo,
        brief,
        manifest_path: manifestPath,
    };

    // Serialise manifest to JSON
    const manifestDict = manifestToDict(manifest);
    manifestDict._metadata = metadata;

    writeFileSync(manifestPath, JSON.stringify(manifestDict, null, 2));

    const algo      = new ManifestAlgo(manifest);
    algo.metadata   = metadata;
    return algo;
}

/**
 * Serialise an AlgoManifest to a plain dict.
 */
function manifestToDict(manifest) {
    return {
        schema_version: manifest.schemaVersion,
        name:           manifest.name,
        description:    manifest.description,
        rationale:      manifest.rationale,
        timeframes:     manifest.timeframes,
        symbols:        manifest.symbols,
        indicators: manifest.indicators.map(i => ({
            id:     i.id,
            kind:   i.kind,
            params: i.params,
        })),
        entries: manifest.entries.map(e => ({
            side: e.side,
            when: e.when,
        })),
        exits: manifest.exits.map(x => {
            const obj = { side: x.side };
            if (x.when   !== null) obj.when   = x.when;
            if (x.slAtr  !== null) obj.sl_atr = x.slAtr;
            if (x.tpAtr  !== null) obj.tp_atr = x.tpAtr;
            return obj;
        }),
        risk: {
            size:           manifest.risk.size,
            atr_mult:       manifest.risk.atrMult,
            fixed_lots:     manifest.risk.fixedLots,
            max_concurrent: manifest.risk.maxConcurrent,
            hedge:          manifest.risk.hedge,
            cool_down_bars: manifest.risk.coolDownBars,
        },
        code: manifest.code,
    };
}

File: cpp/src/algo_gen/promote.cpp
/**
 * AlgoForge — src/algo_gen/promote.cpp
 * S1 Phase 1 — Registry promotion.
 *
 * Mirrors python/algoforge/algo_gen/promote.py:
 *   - Refuses red tier.
 *   - Requires confirm=true for orange/yellow.
 *   - Writes registry_dir/<name>.json with _metadata block.
 */

#include "algo_gen_internal.hpp"
#include "core/algo_gen.hpp"

#include <chrono>
#include <cstdio>
#include <ctime>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <string>

namespace algoforge::algo_gen {

/* =========================================================================
 * Build ISO-8601 UTC timestamp (YYYY-MM-DDTHH:MM:SSZ)
 * ========================================================================= */

static std::string utc_now_iso8601() {
    auto now   = std::chrono::system_clock::now();
    std::time_t t = std::chrono::system_clock::to_time_t(now);
    std::tm tm_utc;
#if defined(_WIN32)
    gmtime_s(&tm_utc, &t);
#else
    gmtime_r(&t, &tm_utc);
#endif
    char buf[32];
    snprintf(buf, sizeof(buf), "%04d-%02d-%02dT%02d:%02d:%02dZ",
             tm_utc.tm_year + 1900,
             tm_utc.tm_mon  + 1,
             tm_utc.tm_mday,
             tm_utc.tm_hour,
             tm_utc.tm_min,
             tm_utc.tm_sec);
    return buf;
}

/* =========================================================================
 * Serialise an AlgoManifest to a JSON string (manifest body only)
 * Used by promote to write the registry file.
 * ========================================================================= */

static std::string manifest_to_json(const AlgoManifest& m) {
    using namespace json;
    std::ostringstream o;
    o << "{\n";
    o << "  \"schema_version\": " << emit_string_val(m.schema_version) << ",\n";
    o << "  \"name\": "           << emit_string_val(m.name)           << ",\n";
    o << "  \"description\": "    << emit_string_val(m.description)    << ",\n";
    o << "  \"rationale\": "      << emit_string_val(m.rationale)      << ",\n";

    // timeframes
    o << "  \"timeframes\": [";
    for (size_t i = 0; i < m.timeframes.size(); ++i) {
        if (i) o << ", ";
        o << emit_string_val(m.timeframes[i]);
    }
    o << "],\n";

    // symbols
    o << "  \"symbols\": ";
    if (std::holds_alternative<std::string>(m.symbols)) {
        o << emit_string_val(std::get<std::string>(m.symbols));
    } else {
        const auto& syms = std::get<std::vector<std::string>>(m.symbols);
        o << "[";
        for (size_t i = 0; i < syms.size(); ++i) {
            if (i) o << ", ";
            o << emit_string_val(syms[i]);
        }
        o << "]";
    }
    o << ",\n";

    // indicators
    o << "  \"indicators\": [\n";
    for (size_t i = 0; i < m.indicators.size(); ++i) {
        const auto& ind = m.indicators[i];
        o << "    {\"id\": " << emit_string_val(ind.id)
          << ", \"kind\": "   << emit_string_val(ind.kind)
          << ", \"params\": {";
        bool first_p = true;
        for (const auto& [k, v] : ind.params) {
            if (!first_p) o << ", ";
            o << emit_string_val(k) << ": " << emit_double(v);
            first_p = false;
        }
        o << "}}";
        if (i + 1 < m.indicators.size()) o << ",";
        o << "\n";
    }
    o << "  ],\n";

    // entries
    o << "  \"entries\": [\n";
    for (size_t i = 0; i < m.entries.size(); ++i) {
        const auto& e = m.entries[i];
        o << "    {\"side\": " << emit_string_val(e.side)
          << ", \"when\": "     << emit_string_val(e.when) << "}";
        if (i + 1 < m.entries.size()) o << ",";
        o << "\n";
    }
    o << "  ],\n";

    // exits
    o << "  \"exits\": [\n";
    for (size_t i = 0; i < m.exits.size(); ++i) {
        const auto& x = m.exits[i];
        o << "    {\"side\": " << emit_string_val(x.side);
        if (x.when)   o << ", \"when\": "   << emit_string_val(*x.when);
        if (x.sl_atr) o << ", \"sl_atr\": " << emit_double(*x.sl_atr);
        if (x.tp_atr) o << ", \"tp_atr\": " << emit_double(*x.tp_atr);
        o << "}";
        if (i + 1 < m.exits.size()) o << ",";
        o << "\n";
    }
    o << "  ],\n";

    // risk
    o << "  \"risk\": {\n";
    o << "    \"size\": "           << emit_string_val(m.risk.size)       << ",\n";
    o << "    \"atr_mult\": "       << emit_double(m.risk.atr_mult)       << ",\n";
    o << "    \"fixed_lots\": "     << emit_double(m.risk.fixed_lots)     << ",\n";
    o << "    \"max_concurrent\": " << m.risk.max_concurrent              << ",\n";
    o << "    \"hedge\": "          << emit_bool(m.risk.hedge)            << ",\n";
    o << "    \"cool_down_bars\": " << m.risk.cool_down_bars              << "\n";
    o << "  }";

    // code (escape-hatch)
    if (m.code) {
        o << ",\n  \"code\": " << emit_string_val(*m.code);
    }

    return o.str();
}

/* =========================================================================
 * Public API: promote
 * ========================================================================= */

void promote(const TierReport& report,
             const AlgoManifest& manifest,
             const std::filesystem::path& registry_dir,
             bool confirm) {
    // Refuse red tier
    if (report.tier == Tier::Red) {
        throw std::runtime_error(
            std::string("Cannot promote algorithm '") + manifest.name +
            "': tier is red (min_score=" +
            [&]{ char b[32]; snprintf(b, sizeof(b), "%.2f", report.min_score); return std::string(b); }() +
            " < 70). Red-tier algorithms are never promoted."
        );
    }

    // Orange/Yellow require manual confirmation
    if ((report.tier == Tier::Orange || report.tier == Tier::Yellow) && !confirm) {
        throw std::runtime_error(
            std::string("Cannot promote '") + manifest.name +
            "' (tier=" + tier_name(report.tier) +
            ") without confirm=true. This tier requires manual review."
        );
    }

    // Create registry directory
    std::error_code ec;
    std::filesystem::create_directories(registry_dir, ec);
    if (ec) {
        throw std::runtime_error(
            "Cannot create registry directory: " + registry_dir.string() +
            ": " + ec.message()
        );
    }

    std::filesystem::path manifest_path = registry_dir / (manifest.name + ".json");

    // Build _metadata block
    using namespace json;
    std::string created_at = utc_now_iso8601();
    std::string tier_str   = tier_name(report.tier);
    std::string color_str  = tier_color(report.tier);

    std::ostringstream meta;
    meta << "  \"_metadata\": {\n";
    meta << "    \"source\": \"algo_gen\",\n";
    meta << "    \"schema_version\": " << emit_string_val(manifest.schema_version) << ",\n";
    meta << "    \"tier\": "           << emit_string_val(tier_str)   << ",\n";
    meta << "    \"tier_color\": "     << emit_string_val(color_str)  << ",\n";
    meta << "    \"scores\": {\n";
    meta << "      \"walk_forward\": " << emit_double(report.walk_forward) << ",\n";
    meta << "      \"mc_bootstrap\": " << emit_double(report.mc_bootstrap) << ",\n";
    meta << "      \"robustness\": "   << emit_double(report.robustness)   << "\n";
    meta << "    },\n";
    meta << "    \"min_score\": "  << emit_double(report.min_score)  << ",\n";
    meta << "    \"size_mult\": "  << emit_double(report.size_mult)  << ",\n";
    meta << "    \"experimental\": " << emit_bool(report.experimental) << ",\n";
    meta << "    \"paper_only\": "   << emit_bool(report.paper_only)   << ",\n";
    meta << "    \"created_at\": "   << emit_string_val(created_at)    << ",\n";
    meta << "    \"generator\": {},\n";
    meta << "    \"brief\": \"\",\n";
    meta << "    \"manifest_path\": " << emit_string_val(manifest_path.string()) << "\n";
    meta << "  }\n";

    // Compose full JSON: manifest body + _metadata
    std::string body = manifest_to_json(manifest);

    std::ofstream f(manifest_path);
    if (!f.is_open()) {
        throw std::runtime_error(
            "Cannot open file for writing: " + manifest_path.string()
        );
    }
    f << body << ",\n" << meta.str() << "}\n";
    f.close();
}

} /* namespace algoforge::algo_gen */
